File: api/services/carbon_service.py
EMISSION_FACTORS = {
    'car_km': 0.192,
    'public_transport_km': 0.041,
    'flight_hour': 250,
    'electricity_kwh': 0.85,
    'ac_hour': 1.5,
}

DIET_MULTIPLIERS = {
    'vegan': 1000,
    'vegetarian': 1500,
    'pescatarian': 1700,
    'mixed': 2500,
    'meat-heavy': 3300,
}

RECYCLING_MULTIPLIERS = {
    'always': 0.5,
    'sometimes': 0.75,
    'rarely': 0.9,
    'never': 1.0,
}

def calculate_carbon(input_data: dict) -> dict:
    transport_data = input_data.get('transportation', {})
    energy_data = input_data.get('energy', {})
    food_data = input_data.get('food', {})
    waste_data = input_data.get('waste', {})

    # 1. Calculate Transportation Emissions (Yearly)
    transport_emissions = (
        (transport_data.get('carDistanceKm', 0) * 12 * EMISSION_FACTORS['car_km']) +
        (transport_data.get('publicTransportKm', 0) * 12 * EMISSION_FACTORS['public_transport_km']) +
        (transport_data.get('flightHours', 0) * EMISSION_FACTORS['flight_hour'])
    )

    # 2. Calculate Energy Emissions (Yearly)
    energy_emissions = (
        (energy_data.get('electricityKwh', 0) * 12 * EMISSION_FACTORS['electricity_kwh']) +
        (energy_data.get('acUsageHours', 0) * 365 * EMISSION_FACTORS['ac_hour'])
    )

    # 3. Calculate Food Emissions (Yearly)
    diet_type = food_data.get('dietType', 'mixed')
    food_emissions = DIET_MULTIPLIERS.get(diet_type, DIET_MULTIPLIERS['mixed'])

    # 4. Calculate Waste Emissions (Yearly)
    waste_generation = waste_data.get('wasteGenerationKg', 0)
    recycling_habit = waste_data.get('recyclingHabit', 'never')
    base_waste_emissions = waste_generation * 52 * 0.5
    waste_emissions = base_waste_emissions * RECYCLING_MULTIPLIERS.get(recycling_habit, 1.0)

    total_emissions = transport_emissions + energy_emissions + food_emissions + waste_emissions

    def get_percentage(value):
        return (value / total_emissions) * 100 if total_emissions > 0 else 0

    max_transport = 10000
    max_energy = 8000
    max_food = 4000
    max_waste = 2000

    def calculate_score(emissions, max_val):
        score = 100 - ((emissions / max_val) * 100)
        return max(0, min(100, round(score)))

    transport_score = calculate_score(transport_emissions, max_transport)
    energy_score = calculate_score(energy_emissions, max_energy)
    food_score = calculate_score(food_emissions, max_food)
    waste_score = calculate_score(waste_emissions, max_waste)

    overall_score = round((transport_score * 0.35) + (energy_score * 0.35) + (food_score * 0.20) + (waste_score * 0.10))

    categories_map = {
        'transportation': transport_emissions,
        'energy': energy_emissions,
        'food': food_emissions,
        'waste': waste_emissions
    }

    highest_category = max(categories_map, key=categories_map.get)

    return {
        'totalEmissionsKgCO2': round(total_emissions),
        'overallScore': overall_score,
        'highestEmissionCategory': highest_category,
        'categories': {
            'transportation': {
                'emissionsKgCO2': round(transport_emissions),
                'percentage': round(get_percentage(transport_emissions)),
                'score': transport_score
            },
            'energy': {
                'emissionsKgCO2': round(energy_emissions),
                'percentage': round(get_percentage(energy_emissions)),
                'score': energy_score
            },
            'food': {
                'emissionsKgCO2': round(food_emissions),
                'percentage': round(get_percentage(food_emissions)),
                'score': food_score
            },
            'waste': {
                'emissionsKgCO2': round(waste_emissions),
                'percentage': round(get_percentage(waste_emissions)),
                'score': waste_score
            }
        }
    }

File: backend/src/services/carbon.service.ts
import { CarbonInput, CarbonResult } from '../types';

// Constants for emission factors (simplified for calculation purposes)
const EMISSION_FACTORS = {
  car_km: 0.192, // kg CO2 per km
  public_transport_km: 0.041, // kg CO2 per km
  flight_hour: 250, // kg CO2 per hour of flight
  electricity_kwh: 0.85, // kg CO2 per kWh (depends on grid, using average)
  ac_hour: 1.5, // kg CO2 per hour of AC
};

const DIET_MULTIPLIERS: Record<string, number> = {
  vegan: 1000, // kg CO2 per year
  vegetarian: 1500,
  pescatarian: 1700,
  mixed: 2500,
  'meat-heavy': 3300,
};

const RECYCLING_MULTIPLIERS: Record<string, number> = {
  always: 0.5, // 50% reduction
  sometimes: 0.75,
  rarely: 0.9,
  never: 1.0,
};

export const calculateCarbon = (input: CarbonInput): CarbonResult => {
  // 1. Calculate Transportation Emissions (Yearly)
  const transportEmissions = 
    (input.transportation.carDistanceKm * 12 * EMISSION_FACTORS.car_km) +
    (input.transportation.publicTransportKm * 12 * EMISSION_FACTORS.public_transport_km) +
    (input.transportation.flightHours * EMISSION_FACTORS.flight_hour);

  // 2. Calculate Energy Emissions (Yearly)
  const energyEmissions = 
    (input.energy.electricityKwh * 12 * EMISSION_FACTORS.electricity_kwh) +
    (input.energy.acUsageHours * 365 * EMISSION_FACTORS.ac_hour);

  // 3. Calculate Food Emissions (Yearly)
  const foodEmissions = DIET_MULTIPLIERS[input.food.dietType] || DIET_MULTIPLIERS.mixed;

  // 4. Calculate Waste Emissions (Yearly)
  // Base waste emission factor roughly 0.5 kg CO2 per kg of waste
  const baseWasteEmissions = input.waste.wasteGenerationKg * 52 * 0.5;
  const wasteEmissions = baseWasteEmissions * (RECYCLING_MULTIPLIERS[input.waste.recyclingHabit] || 1.0);

  const totalEmissionsKgCO2 = transportEmissions + energyEmissions + foodEmissions + wasteEmissions;

  // Determine percentages
  const getPercentage = (value: number) => totalEmissionsKgCO2 > 0 ? (value / totalEmissionsKgCO2) * 100 : 0;

  // Calculate scores (0-100, 100 is best)
  // Arbitrary baseline for max emissions per category to map to 0 score
  const maxTransport = 10000;
  const maxEnergy = 8000;
  const maxFood = 4000;
  const maxWaste = 2000;

  const calculateScore = (emissions: number, max: number) => {
    const score = 100 - ((emissions / max) * 100);
    return Math.max(0, Math.min(100, Math.round(score)));
  };

  const transportScore = calculateScore(transportEmissions, maxTransport);
  const energyScore = calculateScore(energyEmissions, maxEnergy);
  const foodScore = calculateScore(foodEmissions, maxFood);
  const wasteScore = calculateScore(wasteEmissions, maxWaste);

  const overallScore = Math.round((transportScore * 0.35) + (energyScore * 0.35) + (foodScore * 0.20) + (wasteScore * 0.10));

  // Determine highest emission category
  const categoriesMap = {
    transportation: transportEmissions,
    energy: energyEmissions,
    food: foodEmissions,
    waste: wasteEmissions
  };

  const highestEmissionCategory = Object.keys(categoriesMap).reduce((a, b) => 
    categoriesMap[a as keyof typeof categoriesMap] > categoriesMap[b as keyof typeof categoriesMap] ? a : b
  ) as 'transportation' | 'energy' | 'food' | 'waste';

  return {
    totalEmissionsKgCO2: Math.round(totalEmissionsKgCO2),
    overallScore,
    highestEmissionCategory,
    categories: {
      transportation: {
        emissionsKgCO2: Math.round(transportEmissions),
        percentage: Math.round(getPercentage(transportEmissions)),
        score: transportScore
      },
      energy: {
        emissionsKgCO2: Math.round(energyEmissions),
        percentage: Math.round(getPercentage(energyEmissions)),
        score: energyScore
      },
      food: {
        emissionsKgCO2: Math.round(foodEmissions),
        percentage: Math.round(getPercentage(foodEmissions)),
        score: foodScore
      },
      waste: {
        emissionsKgCO2: Math.round(wasteEmissions),
        percentage: Math.round(getPercentage(wasteEmissions)),
        score: wasteScore
      }
    }
  };
};
